### ISSUE TITLE
UnresolvedElementException: Discovered unresolved method during parsing: char[].clone()

### ISSUE THREAD

**Issue by starksm64** [oracle/graal#877]
I have run across an unexpected UnresolvedElementException for char[].clone() while testing native-image generation. I have distilled this down to a trivial example using the 3 attached java classes found in the Archive.zip attachement:
HelloClone.java
IHello.java
RawMsg.java

```
Scotts-iMacPro:getting-started-secure-native starksm$ $GRAALVM_HOME/bin/native-image --verbose HelloClone
Build on Server(pid: 4025, port: 52179)
SendBuildRequest [
-task=com.oracle.svm.hosted.NativeImageGeneratorRunner
-imagecp
/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/svm/builder/svm.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/svm/builder/objectfile.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/svm/builder/pointsto.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/jvmci/graal.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/jvmci/graal-management.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/jvmci/jvmci-api.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/jvmci/jvmci-hotspot.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/boot/graaljs-scriptengine.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/boot/graal-sdk.jar:/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/svm/library-support.jar:/Users/starksm/Dev/JBoss/Protean/quickstarts/getting-started-secure-native
-H:Path=/Users/starksm/Dev/JBoss/Protean/quickstarts/getting-started-secure-native
-H:CLibraryPath=/Users/starksm/Dev/JBoss/Protean/graalvm-ce-1.0.0-rc10/Contents/Home/jre/lib/svm/clibraries/darwin-amd64
-H:Class=HelloClone
-H:Name=helloclone
]
[helloclone:4025]    classlist:     157.88 ms
[helloclone:4025]        (cap):     638.25 ms
[helloclone:4025]        setup:     902.50 ms
[helloclone:4025]     analysis:   1,391.89 ms
error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: char[].clone(). To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Detailed message:
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved method during parsing: char[].clone(). To diagnose the issue you can use the --allow-incomplete-classpath option. The missing method is then reported at run time when it is accessed the first time.
Trace: 
	at parsing IHello.createHello(IHello.java:3)
Call path from entry point to IHello.createHello(char[]): 
	at IHello.createHello(IHello.java:3)
	at HelloClone.main(HelloClone.java:5)
	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:164)
	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

Error: Processing image build request failed
```

Note that if I try to simplify this to simply the following:

```java
public class HelloClone2 {
	public static void main(String[] args) {
		String msg = "Hello world clone";
		String hello = new String(msg.toCharArray().clone());
		System.out.println(hello);
	}
}
```

Then the native-image generation succeeds.

[Archive.zip](https://github.com/oracle/graal/files/2702025/Archive.zip)

**Comment by bobmcwhirter** [comment]
@starksm64 Can you confirm if this is still an issue?

Building with my JDK11-ish Substrate from HEAD, stuff seems to work.

```
[helloclone:27909]    classlist:   1,590.11 ms
[helloclone:27909]        (cap):   1,406.26 ms
[helloclone:27909]        setup:   2,890.43 ms
[helloclone:27909]   (typeflow):   6,354.30 ms
[helloclone:27909]    (objects):   7,467.19 ms
[helloclone:27909]   (features):     389.92 ms
[helloclone:27909]     analysis:  14,480.72 ms
[helloclone:27909]     (clinit):     248.60 ms
[helloclone:27909]     universe:     612.81 ms
[helloclone:27909]      (parse):   1,145.20 ms
[helloclone:27909]     (inline):   4,156.04 ms
[helloclone:27909]    (compile):  11,641.86 ms
[helloclone:27909]      compile:  17,740.75 ms
[helloclone:27909]        image:   2,711.86 ms
[helloclone:27909]        write:     442.62 ms
[helloclone:27909]      [total]:  40,654.58 ms
$ ./helloclone
Hello world clone
```

**Comment by cstancu** [comment]
@bobmcwhirter thanks for double-checking. I'll go ahead and close this. @starksm64 please reopen if this is still an issue.

**Comment by dmlloyd** [comment]
It looks like this is not completely fixed; see https://github.com/quarkusio/quarkus/issues/5113

**Comment by jaikiran** [comment]
@cstancu, I think this might have to be reopened and looked into. My experiments with the application attached by Scott in this thread have shown the following:

1. Running the attached app using `19.2.1` of Graal/Native-image passes.
2. Running the attached app using `19.3.0 Java 11` version of Graal/Native-image passes
3. Running the attached app using `19.3.0 Java 8` version of Graal/Native-image *FAILS* with that `char[].clone()` error message

**Comment by cstancu** [comment]
@dougxc This seems to be a JVMCI problem. The issue is with the call to `msg.clone()` in:
```
public interface IHello {
	static IHello createHello(char[] msg) {
		return new RawMsg(msg.clone());
	}
}
```

`HotSpotConstantPool.lookupMethod()` throws an `UnresolvedJavaMethod` because the call to `CompilerToVM.compilerToVM().lookupMethodInPool(this, index, (byte)opcode)` returns `null`. This is only a JDK 8 problem, on JDK 11 `HotSpotMethod<Object.clone()>` is returned. I tried this with the latest jvmci-19.3-b06.

**Comment by dougxc** [comment]
Thanks for the notice. Looking into this now and should have a fix for 19.3.1.

**Comment by jaikiran** [comment]
@dougxc, I dont know how much this will help but when I debugged this last weekend I noticed the same thing that @cstancu noted in his comment. The reason why it seemed to return null appears to be a 'IllegalAccessError' that was getting thrown with a message along the lines "Illegal access to Object.clone(Object)" which seemed to be originating from some native code. Unfortunately, I didnt have enough time to dig into it more and I wont have access to graal source at least for a couple of days more. Hope this is of any help
